kafkit/registry/sansio.py

```diff
@@ -0,0 +1,795 @@
+            url_vars={'subject': subject, 'version': str(version)})
+
+        schema = fastavro.parse_schema(json.loads(result['schema']))
+
+        try:
+            self.subject_cache.insert(
+                result['subject'],
+                result['version'],
+                schema_id=result['id'],
+                schema=schema)
+        except TypeError:
+            # Can't cache versions like "latest"
+            pass
+
+        return {
+            'id': result['id'],
+            'version': result['version'],
+            'subject': result['subject'],
+            'schema': schema
+        }
+
+
+class MockRegistryApi(RegistryApi):
+    """A mock implementation of the RegistryApi client that doensn't do
+    network operations and provides attributes for introspection.
+    """
+
+    DEFAULT_HEADERS = {
+        'content-type': "application/vnd.schemaregistry.v1+json"
+    }
+
+    def __init__(self, url='http://registry:8081',
+                 status_code=200, headers=None, body=b''):
+        super().__init__(url=url)
+        self.response_code = status_code
+        self.response_headers = headers if headers else self.DEFAULT_HEADERS
+        self.response_body = body
+
+    async def _request(self, method, url, headers, body):
+        self.method = method
+        self.url = url
+        self.headers = headers
+        self.body = body
+        response_headers = self.response_headers.copy()
+        return self.response_code, response_headers, self.response_body
+
+
+class SchemaCache:
+    """A cache of schemas that maintains a mapping of schemas and their IDs
+    in a Schema Registry.
+
+    Notes
+    -----
+    Use key access to obtain the schema either by ID, or by the value of the
+    schema itself.
+    """
+
+    def __init__(self):
+        self._id_to_schema = {}
+        self._schema_to_id = {}
+
+    def insert(self, schema, schema_id):
+        """Insert a schema into the cache.
+
+        Parameters
+        ----------
+        schema : `dict`
+            An Avro schema.
+        schema_id : `int`
+            ID of the schema in a Schema Registry.
+        """
+        # ensure the cached schemas are always parsed, and then serialize
+        # so it's hashable
+        serialized_schema = SchemaCache._serialize_schema(schema)
+
+        self._id_to_schema[schema_id] = serialized_schema
+        self._schema_to_id[serialized_schema] = schema_id
+
+    def __getitem__(self, key):
+        if isinstance(key, int):
+            return json.loads(self._id_to_schema[key])
+        else:
+            # Key must be a schema
+            # Always ensure the schema is parsed
+            schema = key.copy()
+            try:
+                serialized_schema = SchemaCache._serialize_schema(schema)
+            except Exception:
+                # If the schema couldn't be parsed, its not going to be a
+                # valid key anyhow.
+                raise KeyError
```

Comment:
You may want to print out the serialization error, or at least some message with this, as it might be hard for the user to tell the difference between a key not existing, and maybe when there's something troubling in the cache?

Reply:
It looks like the schema registry catches keyerror and then just continues as if it isn't in the cache, and might add it back in or something?

Reply:
Good idea. The client should recover mostly by doing the actual HTTP request, but it doesn't seem like there's any harm forming an error message for anyone who's using the cache directly.